proposals/0304-structured-concurrency.md

```diff
@@ -719,79 +910,34 @@ func download(url: URL) async throws -> Data? {
 }
 ```
 
-
-
-#### Cancellation
-
-It is possible to query for cancellation from within a synchronous task, e.g. while iterating over a loop and wanting to check if we should abort its execution by using the static `Task.isCancelled` property:
-
-```swift
-extension Task { 
-
-  /// Returns `true` if the task is cancelled, and should stop executing.
-  ///
-  /// Always returns `false` when called from code not currently running inside of a `Task`.
-  static var isCancelled: Bool { get }
-  
-  /// Check if the task is cancelled and throw an `CancellationError` if it was.
-  ///
-  /// It is intentional that no information is passed to the task about why it
-  /// was cancelled. A task may be cancelled for many reasons, and additional
-  /// reasons may accrue / after the initial cancellation (for example, if the
-  /// task fails to immediately exit, it may pass a deadline).
-  ///
-  /// The goal of cancellation is to allow tasks to be cancelled in a
-  /// lightweight way, not to be a secondary method of inter-task communication.
-  ///
-  /// Never throws if invoked from code not currently running inside of a `Task`.
-  static func checkCancellation() throws
-}
-```
-
-These work the same as their instance counter parts, except that if invoked from a context that has no Task available, e.g. if invoked from outside of Swift's concurrency model (e.g. directly from a pthread) a default value is returned.
-
-The `isCancelled` property is implemented as:
+We also offer an asynchronous sleep function, which accepts the number of nanoseconds to suspend for:
 
 ```swift
-extension Task {
-  static var isCancelled: Bool { 
-    Task.current?.isCancelled ?? false
-  }
+extension Task where Success == Never, Failure == Never {
+  public static func sleep(_ duration: UInt64) async { ... }
 }
 ```
 
-Which makes sense, because if not executing within a task, such code can never "be cancelled" using Swift's task infrastructure.
-
-This static `isCancelled` property is always safe to invoke, i.e. it may be invoked from synchronous or asynchronous functions and will always return the expected result. Do note however that checking cancellation while concurrently setting cancellation may be slightly racy, i.e. if the `cancel` is performed from another thread, the `isCancelled`
-
-#### Task priorities
-
-Similarly, a static `currentPriority` property is available to check the priority of the currently executing task:
-
-```swift
-extension Task { 
-  static var currentPriority: Task.Priority { 
-    Task.current?.priority ?? Task.Priority.default
-  }
-}
-```
+The sleep function accepts a plain integer as nanoseconds to sleep for which mirrors known top-level functions performing the same action in the synchronous world. Because use-sites look quite explicit in the way they have to previous this call with an `await` keyword (`await Task.sleep(nanos)`), we prefer to use the well-known `sleep` word rather than introduce new words for this functionality.
```

Comment:
[`Darwin.sleep(_:)`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sleep.html) has a `seconds: CUnsignedInt` parameter. Could the proposal specify which other known top-level functions accept nanoseconds?

Using `public static func sleep(forTimeInterval seconds: Double) async` would fit in with the [existing](https://developer.apple.com/documentation/foundation/thread/1413673-sleep) Foundation API.

Typo: "previous" → "prefix"

```suggestion
The sleep function accepts a plain integer as nanoseconds to sleep for which mirrors known top-level functions performing the same action in the synchronous world. Because use-sites look quite explicit in the way they have to prefix this call with an `await` keyword (`await Task.sleep(nanos)`), we prefer to use the well-known `sleep` word rather than introduce new words for this functionality.
```

Reply:
Let's fix the typo here and discuss the `sleep` API as part of the next review.